[RayTune] Output insufficent resource warning msg when autoscaler is …

xwjiang2010 opened the pull request:
…not running.

RayTune currently does not receive a difinitive signal from resource management
about whether a certain request is not fulfilled because of other competing
requests or would never be fulfilled due to resource limitations. As a result,
user complains repeated PENDING status of trials without making any progress.

This implementation is at best a calculated investment to collect some low
hanging fruits.

A proper fix should involve API changes in resource management in the future.

<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

xwjiang2010 commented:
Putting up an initial draft to collect some feedback.

xwjiang2010 commented on python/ray/tune/trial_executor.py:
```diff
@@ -177,7 +187,19 @@ def force_reconcilation_on_next_step_end(self):
+        all_trials_are_pending = all(
+            trial.status == Trial.PENDING for trial in all_trials)
+        if all_trials_are_pending and not is_ray_cluster() and (
+                trial_runner.iteration +
```
I was hoping to find a EVERY_N equivalent for python logging module but can't find any. Open to suggestions if another API to TrialRunner to get iteration count can be saved :)

krfricke replied:
Maybe record the latest iteration when we raised the message?

krfricke replied:
Please also check if we run in this message very often - does 100 iterations mean we print the message every 10 seconds (because of our 0.1s timeout)?

Yard1 replied:
Perhaps it would be better to use time instead of iterations altogether?

xwjiang2010 replied:
> Please also check if we run in this message very often - does 100 iterations mean we print the message every 10 seconds (because of our 0.1s timeout)?
Kai, what 0.1s timeout you are referring to? Is this it? https://github.com/ray-project/ray/blob/master/python/ray/tune/trial_runner.py#L552. I don't think this code path is even exercised in this case, as it's gated by get_running_trials() returning some trial to proceed (None in our case).
So I think you are right - 100 iterations is probably too frequent. 

To Antoni's point, it's probably better to just use time. And we should pick up something that is meaningful - some time that user is just about to wonder what our code is doing. How about 5 seconds?

Yard1 replied:
5 seconds sounds good. We should probably make it an env var too, so users can change/disable it if they want to

xwjiang2010 replied:
> 5 seconds sounds good. We should probably make it an env var too, so users can change/disable it if they want to

Thinking loud here. If we make it an env variable, wouldn't that mean user needs to explicitly set it in their set up commands to take benefit of this warning msg? I would consider these users already more aware of this issue to the point of even setting this in their yaml file. The general users are still likely to just get stuck without hints about what is happening. Thoughts?

Yard1 replied:
We can have default values for env variables - we'd just set it to 5 seconds

krfricke reviewed (commented):
Generally looks good - let's just see about recording the last time we raised the warning.

Also, it would be good to test this somehow, specifically with the case when we have finished trials already but no new trials can be scheduled (because they might have different resource requirements). I think we can use a unit test here, we don't need an end to end test.

Yard1 commented on python/ray/tune/trial_executor.py:
```diff
@@ -177,7 +187,19 @@ def force_reconcilation_on_next_step_end(self):
+        if all_trials_are_pending and not is_ray_cluster() and (
+                trial_runner.iteration +
+                1) % SHOW_MAYBE_INSUFFICIENT_RESOURCE_WARNING_ITER_DELAY == 0:
+            logger.warning(
```
I think we should be using `warnings.warn` here instead, according to https://docs.python.org/3/howto/logging.html. We are not consistent with it in existing code but it would be good to use best practices in new code

Yard1 replied:
Those are good points. We should keep it as logging, then :+1:

xwjiang2010 commented:
> Generally looks good - let's just see about recording the last time we raised the warning.
> 
> Also, it would be good to test this somehow, specifically with the case when we have finished trials already but no new trials can be scheduled (because they might have different resource requirements). I think we can use a unit test here, we don't need an end to end test.

Good point! Yes, my current logic only considers the case of first trial getting stuck. It should be trivial to cover the case you mentioned here. and +100 to unit tests. Do we have a current test suite that I can piggy back this test on?

xwjiang2010 commented:
> > Generally looks good - let's just see about recording the last time we raised the warning.
> > Also, it would be good to test this somehow, specifically with the case when we have finished trials already but no new trials can be scheduled (because they might have different resource requirements). I think we can use a unit test here, we don't need an end to end test.
> 
> Good point! Yes, my current logic only considers the case of first trial getting stuck. It should be trivial to cover the case you mentioned here. and +100 to unit tests. Do we have a current test suite that I can piggy back this test on?

Actually how about removing is_ray_cluster() all together? Or use a different x seconds threshold for the case with autoscaler? Basically, what kind of set up should I run to say with certain confidence that if none of the PENDING trials are proceeding to RUNNING after x seconds with autoscaler, it's reasonable to output a similar warning msg as well.

richardliaw commented on python/ray/tune/trial_runner.py:
```diff
@@ -350,6 +350,10 @@ def search_alg(self):
         return self._scheduler_alg
 
+    @property
+    def iteration(self):
```
type-hint?

richardliaw commented:
> Actually how about removing is_ray_cluster() all together? Or use a different x seconds threshold for the case with autoscaler? Basically, what kind of set up should I run to say with certain confidence that if none of the PENDING trials are proceeding to RUNNING after x seconds with autoscaler, it's reasonable to output a similar warning msg as well.

I think we should definitely use a different case for `is_ray_cluster()`, but still use that as a decision criteria. Time-based warnings sound good to me.

BTW, the autoscaler also has output when it's scaling up, so it'll be important to design the UX holistically.

xwjiang2010 commented:
> BTW, the autoscaler also has output when it's scaling up, so it'll be important to design the UX holistically.

Yes, I do see those log outputs. How about we say "ignore the following msg if you are still seeing xxx in logs" or it is too cumbersome?

richardliaw commented:
> How about we say "ignore the following msg if you are still seeing xxx in logs" or it is too cumbersome?

I think it's a bit cumbersome/adds to the noise.

xwjiang2010 commented:
> > How about we say "ignore the following msg if you are still seeing xxx in logs" or it is too cumbersome?
> 
> I think it's a bit cumbersome/adds to the noise.

Updated the msg, ptal

xwjiang2010 commented on python/ray/tune/tests/test_ray_trial_executor.py:
```diff
@@ -280,6 +282,59 @@ def create_trial(cpu, gpu=0):
+                    "cpu": 1,
+                    "gpu": 1,
+                })
+            msg = "Autoscaler is disabled. Resource is not ready after " \
```
'/\'

Yard1 replied:
```python
msg = ("Autoscaler is disabled. Resource is not ready after "
             "extended amount of time without any trials running - "
             "please consider if the allocated resource is not enough.")
```

Yard1 commented on python/ray/tune/trial_executor.py:
```diff
@@ -174,10 +192,33 @@ def on_step_end(self, trial_runner):
+                              "this message. " if autoscaler_enabled
+                               else "Autoscaler is disabled. ")
+                logger.warn(
+                    warn_prefix +
+                    "Resource is not ready after extended amount of time "
```
F-strings are preferable
```suggestion
                    f"{warn_prefix}Resource is not ready after extended amount of time "
```

xwjiang2010 replied:
Thanks! Changed to multi-line f-string.

Yard1 commented on python/ray/tune/trial_executor.py:
```diff
@@ -8,6 +10,16 @@
 logger = logging.getLogger(__name__)
 
 
+def get_warn_threshold(autoscaler_enabled: bool):
```
```suggestion
def get_warn_threshold(autoscaler_enabled: bool) -> float:
```

xwjiang2010 commented on python/ray/tune/tests/test_ray_trial_executor.py:
```diff
@@ -280,6 +282,59 @@ def create_trial(cpu, gpu=0):
+            pass
+
+        with self.assertLogs(logger="ray.tune.trial_executor") as ctx:
+            out = tune.run(
```
The set up in this file either drives the tests from `tune.run()` or on a granular level by interacting(start/stop/pause/resume) with individual trials. 

For me, the added logic is more or less hooked up with step(), thus driven by tune.run(). Is there a good way to drive heterogeneous trials? So that we can exercise the logic of:
- one trial runs and terminates successfully
- another trial requires more  resource than the 1st trial and thus we print the warning msg.

I am open to ideas to add more test coverage.

xwjiang2010 replied:
@krfricke to see if any input for unit tests.

xwjiang2010 commented:
> Generally looks good - let's just see about recording the last time we raised the warning.
> 
> Also, it would be good to test this somehow, specifically with the case when we have finished trials already but no new trials can be scheduled (because they might have different resource requirements). I think we can use a unit test here, we don't need an end to end test.

Added tests and left a question of how to have heterogeneous trial running through Tune.run(), so that this specific case you mentioned can also be tested.

xwjiang2010 commented:
Additional questions before I forgot:
- I added some new dependencies, do I need to update our dependency somewhere like requirements.txt?
- I added two env variables, should I update user facing doc somewhere?

Yard1 commented:
@xwjiang2010 
- if it's a Tune test only dependency, add it to `python/requirements/tune/requirements_tune.txt`
- `doc/source/tune/user-guide.rst`, "Environment variables" section

xwjiang2010 commented:
> @xwjiang2010
> 
> * if it's a Tune test only dependency, add it to `python/requirements/tune/requirements_tune.txt`
> * `doc/source/tune/user-guide.rst`, "Environment variables" section

Done! PTAL.

Yard1 commented on doc/source/tune/user-guide.rst:
```diff
@@ -846,6 +846,13 @@ These are the environment variables Ray Tune currently considers:
+  the warning message is printed repeatedly every this amount of time. Defaults to 1 (seconds).
+* **TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S_AUTOSCALER**: Threshold for warning user, when autoscaler is enabled,
+  if no active trials are in RUNNING state for this amount of time.
+  If Tune job is stuck in this state(most likely due to insufficient resource), the warning message is printed
```
```suggestion
  If Tune job is stuck in this state (most likely due to insufficient resource), the warning message is printed
```
missing space

Yard1 commented on doc/source/tune/user-guide.rst:
```diff
@@ -846,6 +846,13 @@ These are the environment variables Ray Tune currently considers:
   be disabled by setting this to lower or equal to 0.
 * **TUNE_WARN_THRESHOLD_S**: Threshold for logging if an Tune event loop operation takes too long. Defaults to 0.5 (seconds).
+* **TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S**: Threshold for warning user if no active trials are in RUNNING state
+  for this amount of time. If Tune job is stuck in this state(most likely due to insufficient resource),
```
```suggestion
  for this amount of time. If Tune job is stuck in this state (most likely due to insufficient resource),
```
missing space

Yard1 commented on python/ray/tune/trial_executor.py:
```diff
@@ -8,6 +10,16 @@
 logger = logging.getLogger(__name__)
 
 
+def get_warn_threshold(autoscaler_enabled: bool) -> float:
```
can we make this private actually? `_get_warn_threshold`

xwjiang2010 replied:
good point. made it private.

Yard1 commented:
Awesome, thanks! Some minor nits and we are good to go

krfricke commented on doc/source/tune/user-guide.rst:
```diff
@@ -846,6 +846,13 @@ These are the environment variables Ray Tune currently considers:
   trial startups. After the grace period, Tune will block until a result from a running trial is received. Can
   be disabled by setting this to lower or equal to 0.
 * **TUNE_WARN_THRESHOLD_S**: Threshold for logging if an Tune event loop operation takes too long. Defaults to 0.5 (seconds).
+* **TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S**: Threshold for warning user if no active trials are in RUNNING state
+  for this amount of time. If Tune job is stuck in this state (most likely due to insufficient resource),
+  the warning message is printed repeatedly every this amount of time. Defaults to 1 (seconds).
+* **TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S_AUTOSCALER**: Threshold for warning user, when autoscaler is enabled,
+  if no active trials are in RUNNING state for this amount of time.
+  If Tune job is stuck in this state (most likely due to insufficient resource), the warning message is printed
+  repeatedly every this amount of time. Defaults to 60 (seconds).
```
```suggestion
* **TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S**: Threshold for throwing a warning if no active trials are in ``RUNNING`` state
  for this amount of seconds. If the Ray Tune job is stuck in this state (most likely due to insufficient resources),
  the warning message is printed repeatedly every this amount of time. Defaults to 1 (seconds).
* **TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S_AUTOSCALER**: Threshold for throwing a warning when the autoscaler is enabled and
  if no active trials are in ``RUNNING`` state for this amount of time.
  If the Ray Tune job is stuck in this state (most likely due to insufficient resources), the warning message is printed
  repeatedly every this amount of time. Defaults to 60 (seconds).
```

krfricke commented on python/ray/tune/trial_executor.py:
```diff
@@ -8,6 +10,16 @@
+                "TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S_AUTOSCALER", "60"))
+    else:
+        return float(
+            os.environ.get("TUNE_WARN_INSUFFICENT_RESOURCE_THRESHOLD_S", "5"))
```
In the docs above the default time is `1` second, so we should sync that

krfricke commented on python/ray/tune/trial_executor.py:
```diff
@@ -174,10 +192,33 @@ def on_step_end(self, trial_runner):
+                               "Autoscaler is disabled.")
+                logger.warning(
+                    f"{warn_prefix} "
+                    f"Resource is not ready after extended amount of time "
+                    f"without any trials running - please consider if the "
+                    f"allocated resource is not enough.")
```
Let's make this a bit more explicit and actionable here, suggestion:

```
No trial is running and no new trial has been started within the last {time} seconds. This could be due to the cluster not having enough resources available to start the next trial. Please check if the requested resources can be fulfilled by your cluster, or will be fulfilled eventually (when using the Ray autoscaler).
```

krfricke commented on python/requirements/tune/requirements_tune.txt:
```diff
@@ -28,6 +29,7 @@ optuna==2.8.0
 pytest-remotedata==0.3.2
 pytorch-lightning-bolts==0.2.5
 pytorch-lightning==1.0.3
+mock==4.0.3
```
can we keep alphabetization here if we add new packages?

xwjiang2010 commented on python/ray/tune/tests/test_ray_trial_executor.py:
```diff
@@ -1,4 +1,6 @@
 # coding: utf-8
+from freezegun import freeze_time
```
This is to mock the behavior of anything related to time.time() or time.monotonic() etc.
https://github.com/spulec/freezegun

I think it's a neat mock to have, especially since we have these time checks here and there in our codebase. A pretty common and popular construct for unit testing - available in multiple languages.

krfricke replied:
Ok, let's keep it then

richardliaw commented on python/ray/tune/trial_executor.py:
```diff
@@ -12,6 +15,18 @@
 logger = logging.getLogger(__name__)
 
 
+# Accessing environment variable could be slow.
```
wait really? do you have some reference for this?

xwjiang2010 replied:
I was under the impression that accessing environment variable incurs penalty for some scripting languages. But looking closely at os.py, it seems just a normal wrapped dictionary in process. So maybe not so much in this case.

richardliaw commented on python/ray/tune/trial_executor.py:
```diff
@@ -194,6 +215,31 @@ def on_step_end(self, trials: List[Trial]) -> None:
+                    f"No trial is running and no new trial has been started "
+                    f"within at least the last "
+                    f"{_get_warn_threshold(autoscaler_enabled)} seconds. "
+                    f"This could be due to the cluster not having enough "
+                    f"resources available to start the next trial. Please "
+                    f"check if the requested resources can be fulfilled by "
+                    f"your cluster, or will be fulfilled eventually (when "
+                    f"using the Ray autoscaler).")
```
A couple comments here:

1. IMO users that aren't using the Ray autoscaler should not see "Ray autoscaler"
2. This doesn't actually provide any action for the user to take. For example, user may not know what "requested resources" means nor even "cluster". 

Instead, it would be good to say:

1. which resource is not available, and how much is being requested
2. what is the total amount of those resource available on the cluster

Also, one suggestion should be to say that they should stop their tuning job and reconfigure their resource request.

Does that make sense? In principle, we should provide 1. what went wrong on the Ray side (in terms that the end user understands) 2. what the user did wrong (if possible) 3. what they should do instead :)

xwjiang2010 replied:
This is good feedback. 
Practically there is no API that exposes those information. Left a TODO and filed https://github.com/ray-project/ray/issues/17799 to follow up.
